src/Utilities/product_of_sets.jl

```diff
@@ -140,86 +146,85 @@ Product of sets in the order the constraints are added, grouping the
 constraints of the same types contiguously.
 """
 abstract type OrderedProductOfSets{T} <: ProductOfSets{T} end
+
 macro product_of_sets(name, set_types...)
     esc_name = esc(name)
     T = esc(:T)
     type_def = :(
-        struct $esc_name{$T} <: $MOIU.OrderedProductOfSets{$T}
+        struct $(esc_name){$(T)} <: $MOIU.OrderedProductOfSets{$(T)}
+            """
+            The number of rows that each set takes.
+            """
             num_rows::Vector{Int}
+
+            """
+            A dictionary which maps the `set_index` and `offset` of a set to the
+            dimension, i.e., `dimension[(set_index,offset)] → dim`.
+            """
             dimension::Dict{Tuple{Int,Int},Int}
-            function $esc_name{$T}() where {$T}
-                return new(zeros(Int, $(1 + length(set_types))), Dict{Int,Int}())
+
+            function $(esc_name){$(T)}() where {$(T)}
+                return new(
+                    zeros(Int, $(length(set_types))),
```

Comment:
@blegat I changed `num_rows` to store the number of elements, rather than the cumulative list. It makes the code simpler, and the extra summations required are unlikely to cause a bottleneck.

Reply:
When there is only scalar set, `rows` is trivial with `num_rows` being the cumulative sum. Having to take a cumulative sum should make it much slower. I don't think the simplification is worth it.

Reply:
> When there is only scalar set

Then they should use `MixOfScalarSets`.

> Having to take a cumulative sum should make it much slower

I don't think so, because this vector is usually pretty small (e.g., length 4). Having to maintain the cumulative vector every time we add a new set also takes time.

Reply:
> Then they should use `MixOfScalarSets`.

Not so sure, `MixOfScalarSets` is slower since it needs one push per constraints so if you don't mind your constraints being reordered, `ProductOfSets` is faster.

> Having to maintain the cumulative vector every time we add a new set also takes time.

We don't have to maintain it every time, we update it at the end of `copy_to`. You cannot add a constraint outside of `copy_to`. During `copy_to` it represents the number for each type of set and at the end of `copy_to`, it represents the cumulative.

Reply:
Why do we have `MixOfScalarSets` then? We should just be aiming to minimize the surface area of this code.

The previous code was mutating `.sets` when it called `MOI.dimension`, and so was implicitly assuming that `MOI.dimension` was called once and only once, and at the beginning of `final_touch` before everything else. That seems very brittle. 

Once we have things working, with solvers hooked up, we can return with benchmarks. Until then, I'm leaving as-is.

Reply:
Then I suggest having a `final_touch` for the sets API as well. This is consistent with `MatrixOfConstraints` and `MutableSparseMatrixCSC`. In `final_touch` you do the cumulative sum and in `dimension` you just return `sets.num_rows[end]`.
If you look at `colptr` of `MutableSparseMatrixCSC`, the meaning of the stored values also changes throughout the API calls. These are not easy to use API that you can safely use in any order. It's only meant to be used by `MatrixOfConstraints` or adventurous users.

> Why do we have `MixOfScalarSets` then?

Not sure yet where we are headed but I wrote them to compare. We need to have that at least even here or in MatrixOptInterface since it's useful for some formats there where the constraints have mixed order.

Reply:
Okay, I added `final_touch` and did the ordering.

Reply:
> useful for some formats there where the constraints have mixed order

Which formats? We can always add things in later, but we don't want to be carrying code and tests just for the sake of it.